cmd/csi-addons/volumegroup.go

```diff
@@ -0,0 +1,281 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/csi-addons/kubernetes-csi-addons/internal/proto"
+	"github.com/csi-addons/kubernetes-csi-addons/internal/sidecar/service"
+)
+
+// VolumeGroupBase struct for common fields
+type VolumeGroupBase struct {
+	volumeGroupID   string
+	volumeGroupName string
+	secretName      string
+	secretNamespace string
+	parameters      map[string]string
+	volumeIDs       []string
+}
+
+// SetVolumeGroupID sets the volumeGroupID and checks if it's empty
+func (vgb *VolumeGroupBase) SetVolumeGroupID(volumeGroupID string) error {
+	if volumeGroupID == "" {
+		return errors.New("volumeGroupID is not set")
+	}
+	vgb.volumeGroupID = volumeGroupID
+	return nil
+}
+
+// SetVolumeGroupName sets the volumeGroupName and checks if it's empty
+func (vgb *VolumeGroupBase) SetVolumeGroupName(volumeGroupName string) error {
+	if volumeGroupName == "" {
+		return errors.New("volumeGroupName is not set")
+	}
+	vgb.volumeGroupName = volumeGroupName
+	return nil
+}
+
+// SetSecret sets the secret name and namespace
+func (vgb *VolumeGroupBase) SetSecret(secret string) error {
+	secretNamespace, secretName, err := parseSecret(secret)
+	if err != nil {
+		return fmt.Errorf("failed to parse secret: %w", err)
+	}
+	vgb.secretNamespace = secretNamespace
+	vgb.secretName = secretName
+	return nil
+}
+
+// SetParameters sets and parses the parameters
+func (vgb *VolumeGroupBase) SetParameters(parameters string) error {
+	params, err := parseParameters(parameters)
+	if err != nil {
+		return fmt.Errorf("failed to parse parameters: %w", err)
+	}
+	vgb.parameters = params
+	return nil
+}
+
+// SetVolumeIDs sets and parses the volume IDs
+func (vgb *VolumeGroupBase) SetVolumeIDs(volumeids string) {
+	volumeIDs := strings.Split(volumeids, ",")
+	vgb.volumeIDs = volumeIDs
+}
+
+// Parses the parameters to convert them from string format to map[string]string format
+func parseParameters(paramString string) (map[string]string, error) {
+	params := make(map[string]string)
+	if paramString != "" {
+		pairs := strings.Split(paramString, ",")
+		for _, pair := range pairs {
+			kv := strings.SplitN(pair, "=", 2)
+			if len(kv) == 2 {
+				key := kv[0]
+				value := kv[1]
+				params[key] = value
+			} else {
+				return nil, fmt.Errorf("invalid parameter : %s", pair)
+			}
+		}
+	}
+	return params, nil
+}
+
+// Splits the secrets passed as namespace/name
+func parseSecret(secretString string) (string, string, error) {
```

Comment:
can you please update other function to use this function to parse the secret?

Reply:
@Madhu-1 `parseSecret` is being used in `SetSecret` which in turn is being used in the other functions. If I have to use `parseSecret` I would have to delete `SetSecret` and use `parseSecret` in all the functions. Is that what is expected?

Reply:
SetSecret is specific to the common struct, if we can use the common struct that would be good, if not lets leave that as it is not use parseSecret and we can implement common struct which is generic for all other operations.